core/modules/user/user.entity.inc

```diff
@@ -371,16 +371,16 @@ class UserStorageController extends EntityDatabaseStorageController {
       }
     }
     else {
-      // Save user roles.
-      if (count($entity->roles) > 1) {
+      // Save user roles. Skip built-in roles, and ones that were already saved
+      // to the database during hook calls.
+      $rids_to_skip = array_merge(array(BACKDROP_ANONYMOUS_ROLE, BACKDROP_AUTHENTICATED_ROLE), db_query('SELECT role FROM {users_roles} WHERE uid = :uid', array(':uid' => $entity->uid))->fetchCol());
+      if ($rids_to_save = array_diff(array_keys($entity->roles), $rids_to_skip)) {
```

Comment:
This is likely the problem with this PR. The list of roles in $user->roles is now an unordered list of role names, rather than indexed by RID.